File: api/src/main/java/me/kvdpxne/dtm/state/MutableState.java
package me.kvdpxne.dtm.state;

/**
 * Represents a mutable state that tracks whether an object has been modified.
 * This interface extends {@link State} and allows for explicitly setting or
 * toggling the modified state of the object.
 * <p>
 * Useful in scenarios where changes to an object need to be tracked and
 * controlled, such as managing persistence layers or undo/redo operations.
 *
 * @since 0.1.0
 */
public interface MutableState
  extends State {

  /**
   * Checks if the object has been modified.
   * <p>
   * This overrides the method from {@link State} to retain its contract while
   * being part of a mutable state interface.
   *
   * @return {@code true} if the object has been modified; {@code false}
   * otherwise.
   * @since 0.1.0
   */
  @Override
  boolean wasModified();

  /**
   * Sets the modified state of the object explicitly.
   *
   * @param modified the new modified state to set.
   * @since 0.1.0
   */
  void setModified(
    final boolean modified
  );

  /**
   * Marks the object as modified.
   * <p>
   * This is a convenience method to set the modified state to {@code true}.
   *
   * @since 0.1.0
   */
  void markAsModified();

  /**
   * Unmarks the object as modified.
   * <p>
   * This is a convenience method to set the modified state to {@code false}.
   *
   * @since 0.1.0
   */
  void unmarkAsModified();
}

File: src/main/kotlin/me/kvdpxne/dtm/data/state/MutableState.kt
package me.kvdpxne.dtm.data.state

/**
 * An interface representing a mutable state that can be marked as modified.
 *
 * Extends the [State] interface, adding functionality to track whether the
 * state has been changed and to reset this modified status as needed.
 *
 * Implementations of this interface are expected to use the [wasModified] flag
 * to indicate when a state has been modified, enabling systems to respond
 * appropriately to changes (e.g., triggering updates, logging changes, etc.).
 *
 * @since 0.1.0
 */
interface MutableState : State {

  /**
   * A flag indicating whether the state has been modified since the last
   * reset.
   *
   * This is useful for tracking changes, especially in contexts where state
   * persistence or synchronization is required.
   *
   * @return `true` if the state was modified; `false` otherwise.
   * @since 0.1.0
   */
  val wasModified: Boolean

  /**
   * Marks the state as modified. This sets [wasModified] to `true`, allowing
   * systems to recognize that a change has occurred.
   *
   * This function should be called whenever a modification is made to the
   * state, ensuring that [wasModified] accurately reflects the current state.
   *
   * @since 0.1.0
   */
  fun markAsModified()

  /**
   * Resets the modification status of the state, setting [wasModified] to
   * `false`.
   *
   * This can be used after the state changes have been acknowledged or
   * processed, allowing the system to monitor for new modifications going
   * forward.
   *
   * @since 0.1.0
   */
  fun unmarkAsModified()
}